src/framework/ui/internal/uiarrangement.cpp

```diff
@@ -0,0 +1,168 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "uiarrangement.h"
+
+#include <QJsonDocument>
+#include <QJsonValue>
+#include <QJsonArray>
+
+#include "log.h"
+
+using namespace mu::ui;
+using namespace mu::workspace;
+
+void UiArrangement::load()
+{
+    workspacesDataProvider()->workspaceChanged().onNotify(this, [this]() {
+        updateData(DataKey::UiSettings, m_settings, m_valuesNotifications);
+        updateData(DataKey::UiStates, m_states, m_statesNotifications);
+        updateData(DataKey::UiToolConfigs, m_toolconfigs, m_toolconfigsNotifications);
+    });
+
+    updateData(DataKey::UiSettings, m_settings, m_valuesNotifications);
+    updateData(DataKey::UiStates, m_states, m_statesNotifications);
+    updateData(DataKey::UiToolConfigs, m_toolconfigs, m_toolconfigsNotifications);
+}
+
+void UiArrangement::updateData(DataKey key, QJsonObject& obj, Notifications& notifications) const
+{
+    RetVal<QByteArray> data = workspacesDataProvider()->rawData(key);
+    if (!data.ret) {
+        LOGD() << "no data: " << key_to_string(key);
+        obj = QJsonObject();
+        return;
+    }
+
+    QJsonObject oldObj = obj;
+    obj = QJsonDocument::fromJson(data.val).object();
+
+    Notifications::Iterator it = notifications.begin();
+    for (; it != notifications.end(); ++it) {
+        const QString& key = it.key();
+
+        if (oldObj.value(key) != obj.value(key)) {
+            it.value().notify();
+        }
+    }
+}
+
+void UiArrangement::saveData(workspace::DataKey key, const QJsonObject& obj)
+{
+    QByteArray data = QJsonDocument(obj).toJson();
+
+    workspacesDataProvider()->setRawData(key, data);
+}
+
+QString UiArrangement::value(const QString& key) const
+{
+    QJsonValue val = m_settings.value(key);
+    return val.toString();
+}
+
+void UiArrangement::setValue(const QString& key, const QString& val)
+{
+    m_settings[key] = val;
+    saveData(DataKey::UiSettings, m_settings);
+    if (m_valuesNotifications.contains(key)) {
+        m_valuesNotifications[key].notify();
+    }
+}
+
+mu::async::Notification UiArrangement::valueChanged(const QString& key) const
+{
+    return m_valuesNotifications[key];
+}
+
+QByteArray UiArrangement::state(const QString& key) const
+{
+    QJsonValue val = m_states.value(key);
+    QString valStr = val.toString();
+    return valStr.toLocal8Bit();
+}
+
+void UiArrangement::setState(const QString& key, const QByteArray& data)
+{
+    m_states[key] = QString::fromLocal8Bit(data);
+    saveData(DataKey::UiStates, m_states);
+    if (m_statesNotifications.contains(key)) {
+        m_statesNotifications[key].notify();
+    }
+}
+
+mu::async::Notification UiArrangement::stateChanged(const QString& key) const
+{
+    return m_statesNotifications[key];
+}
+
+ToolConfig UiArrangement::toolConfig(const QString& toolName) const
+{
+    TRACEFUNC;
+
+    //! NOTE Maybe we need to cache?
```

Comment:
makes sense to me, but not in this PR I guess

Reply:
I added `TRACEFUNC`, we need watch for it values